crates/common/types/block.rs

```diff
@@ -36,16 +36,58 @@ use once_cell::sync::OnceCell;
 pub struct Block {
     pub header: BlockHeader,
     pub body: BlockBody,
+
+    // Cache these to avoid doing repeat calculations, since we need the RLP encoded size when checking for the size limit
+    // We cache the RLP encodes of the body and the header rather than the entire block because these are the ones that are stored
+    #[serde(skip)]
+    #[rkyv(with = Skip)]
+    pub cached_body_rlp_encode: OnceCell<Vec<u8>>,
+    #[serde(skip)]
+    #[rkyv(with = Skip)]
+    pub cached_header_rlp_encode: OnceCell<Vec<u8>>,
 }
 
 impl Block {
     pub fn new(header: BlockHeader, body: BlockBody) -> Block {
-        Block { header, body }
+        Block {
+            header,
+            body,
+            cached_body_rlp_encode: OnceCell::new(),
+            cached_header_rlp_encode: OnceCell::new(),
+        }
     }
 
     pub fn hash(&self) -> BlockHash {
         self.header.hash()
     }
+
+    // Calculate the size of the RLP encode of the block
+    // We need the RLP encode in two places: in the block validation to check its size doesn't exceed the maximum, and in the store to save it
+    // However, the store uses the encode of the body and the header, whereas for the max size we need the size of the encode of the block as a whole
+    // So we cache the body and header, and calculate the size of the block encode based off of them
+    pub fn get_rlp_encode_size(&self) -> usize {
```

Comment:
I would expect the block to call something like block.encode().size(), not have a special function for the size